server/blob/blob-union/src/main/java/org/apache/james/blob/union/HybridBlobStore.java

```diff
@@ -128,45 +145,62 @@ private HybridBlobStore(BlobStore lowCostBlobStore, BlobStore highPerformanceBlo
-    }
-
-    private Mono<BlobStore> selectBlobStore(StoragePolicy storagePolicy, Mono<Boolean> largeData) {
         switch (storagePolicy) {
             case LOW_COST:
-                return Mono.just(lowCostBlobStore);
+                return Mono.from(lowCostBlobStore.save(bucketName, data, storagePolicy))
+                    .doFinally(any -> closeInputStream(data));
+
             case SIZE_BASED:
-                return largeData.map(isLarge -> {
-                    if (isLarge) {
-                        return lowCostBlobStore;
-                    }
-                    return highPerformanceBlobStore;
-                });
-            case HIGH_PERFORMANCE:
-                return Mono.just(highPerformanceBlobStore);
+                BufferedInputStream bufferedInputStream = new BufferedInputStream(data, configuration.getSizeThreshold() + 1);
+                return Mono.fromCallable(() -> isLargeStream(bufferedInputStream))
+                    .filter(Boolean::booleanValue)
+                    .flatMap(ignored -> Mono.from(lowCostBlobStore.save(bucketName, bufferedInputStream, storagePolicy)))
+                    .switchIfEmpty(checkingDuplicateWriteThenSave(bucketName, readAllBytes(bufferedInputStream), storagePolicy))
+                    .doFinally(any -> closeInputStream(data));
+
             default:
                 throw new RuntimeException("Unknown storage policy: " + storagePolicy);
         }
     }
 
-    private boolean isItABigStream(InputStream bufferedData) throws IOException {
+    private void closeInputStream(InputStream inputStream) {
+        try {
+            inputStream.close();
+        } catch (IOException e) {
+            LOGGER.error(e.getMessage(), e);
+        }
+    }
+
+    private Mono<BlobId> checkingDuplicateWriteThenSave(BucketName bucketName, Mono<byte[]> bytesData, StoragePolicy storagePolicy) {
+        return Mono.fromCallable(configuration::isDuplicateWrite)
+            .filter(Boolean::booleanValue)
+            .flatMap(allowDuplicateWrite ->
+                bytesData.flatMap(bytes ->
+                    Mono.from(lowCostBlobStore.save(bucketName, bytes, storagePolicy))
+                        .then(Mono.from(highPerformanceBlobStore.save(bucketName, bytes, storagePolicy)))))
+            .switchIfEmpty(bytesData.flatMap(bytes -> Mono.from(highPerformanceBlobStore.save(bucketName, bytes, storagePolicy))));
+    }
+
+    private boolean isLargeStream(BufferedInputStream bufferedData) throws IOException {
         bufferedData.mark(0);
         bufferedData.skip(configuration.getSizeThreshold());
         boolean isItABigStream = bufferedData.read() != -1;
         bufferedData.reset();
         return isItABigStream;
     }
 
+    private Mono<byte[]> readAllBytes(BufferedInputStream bufferedInputStream) {
+        // In this case: InputStream are small and we need to reused data from InputStream
```

Comment:
The comment should be placed in the caller of this method, here you have no idea as these conditions depends on the caller.